### ISSUE TITLE
Typensense 0.25.1 on HA cluster and as leader crash 

### ISSUE THREAD

**Issue by bonsempregos** [typesense/typesense#1269]
## Description

I have an issue with typesense on HA cluster as nos being leader crash. This was on normal operation, no several load increase...

## Steps to reproduce

I only have the stack traces from typesense and syslog

syslog:

```
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360119 202086 backward.hpp:4200] Stack trace (most recent call last) in thread 202086:
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360256 202086 backward.hpp:4200] #11   Object "", at 0xffffffffffffffff, in
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360287 202086 backward.hpp:4200] #10   Object "/usr/lib/x86_64-linux-gnu/libc.so.6", at 0x7f99b9f109ff, in
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360307 202086 backward.hpp:4200] #9    Object "/usr/lib/x86_64-linux-gnu/libc.so.6", at 0x7f99b9e7eb42, in
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360332 202086 backward.hpp:4200] #8    Object "/usr/bin/typesense-server", at 0x55ff817a85c3, in execute_native_thread_routine
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360354 202086 backward.hpp:4200] #7  | Source "include/threadpool.h", line 57, in operator()
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360373 202086 backward.hpp:4200]       Source "/usr/include/c++/10/future", line 1592, in ThreadPool [0x55ff7f64f23c]
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360390 202086 backward.hpp:4200] #6  | Source "/usr/include/c++/10/future", line 1459, in _M_set_result
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360416 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/future", line 412, in call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>()>*, bool*), std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>()>*, bool*>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360440 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/mutex", line 729, in __gthread_once
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360457 202086 backward.hpp:4200]       Source "/usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h", line 700, in _M_run [0x55ff7f73f0c3]
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360479 202086 backward.hpp:4200] #5    Object "/usr/lib/x86_64-linux-gnu/libc.so.6", at 0x7f99b9e83f67, in
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360503 202086 backward.hpp:4200] #4  | Source "/usr/include/c++/10/future", line 572, in operator()
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360522 202086 backward.hpp:4200]       Source "/usr/include/c++/10/bits/std_function.h", line 622, in _M_do_set [0x55ff7f64e062]
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360546 202086 backward.hpp:4200] #3  | Source "/usr/include/c++/10/bits/std_function.h", line 292, in __invoke_r<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<_Fn, _Alloc, _Res(_Args ...)>::_M_run<std::_Bind<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>()>, std::allocator<int>, void, {}>::<lambda()>, void>&>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360572 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 115, in __invoke_impl<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<_Fn, _Alloc, _Res(_Args ...)>::_M_run<std::_Bind<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>()>, std::allocator<int>, void, {}>::<lambda()>, void>&>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360622 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 60, in operator()
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360648 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/future", line 1397, in operator()
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360672 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/future", line 1456, in __invoke_r<void, std::_Bind<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>()>&>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360692 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 110, in __invoke_impl<void, std::_Bind<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>()>&>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360713 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 60, in operator()<>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360731 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/functional", line 499, in __call<void>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360749 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/functional", line 416, in __invoke<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>&>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360770 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 95, in __invoke_impl<void, HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>&>
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360793 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 60, in operator()
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360811 202086 backward.hpp:4200]       Source "src/http_server.cpp", line 684, in _M_invoke [0x55ff7f740e23]
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360832 202086 backward.hpp:4200] #2    Source "src/core_api.cpp", line 403, in get_search [0x55ff7f7075cc]
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360855 202086 backward.hpp:4200] #1  | Source "external/lrucache/include/lru/internal/base-cache.hpp", line 913, in _insert_new_key
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360872 202086 backward.hpp:4200]     | Source "external/lrucache/include/lru/internal/base-cache.hpp", line 1544, in _evict_lru_for
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360894 202086 backward.hpp:4200]       Source "external/lrucache/include/lru/internal/base-cache.hpp", line 1558, in insert [0x55ff7f71bd26]
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:02.360916 202086 backward.hpp:4200] #0    Object "/usr/bin/typesense-server", at 0x55ff817292d1, in std::__detail::_List_node_base::_M_transfer(std::__detail::_List_node_base*, std::__detail::_List_node_base*)
Oct  2 20:46:02 XXXXX-HOST-XXXXX typesense-server[201933]: Segmentation fault (Address not mapped to object [(nil)])
Oct  2 20:46:09 XXXXX-HOST-XXXXX typesense-server[201933]: E20231002 20:46:09.718156 202086 typesense_server.cpp:121] Typesense 0.25.1 is terminating abruptly.
Oct  2 20:46:10 XXXXX-HOST-XXXXX systemd[1]: typesense-server.service: Main process exited, code=killed, status=11/SEGV
Oct  2 20:46:10 XXXXX-HOST-XXXXX systemd[1]: typesense-server.service: Failed with result 'signal'.
Oct  2 20:46:10 XXXXX-HOST-XXXXX systemd[1]: typesense-server.service: Consumed 21h 49min 49.529s CPU time.
Oct  2 20:46:10 XXXXX-HOST-XXXXX systemd[1]: typesense-server.service: Scheduled restart job, restart counter is at 1.
Oct  2 20:46:10 XXXXX-HOST-XXXXX systemd[1]: typesense-server.service: Consumed 21h 49min 49.529s CPU time.
Oct  2 20:46:10 XXXXX-HOST-XXXXX typesense-server[2106225]: Log directory is configured as: /logs
```

typesense log

```
I20231002 20:45:55.868680 202400 raft_server.cpp:564] Term: 4, last_index index: 95455, committed_index: 95455, known_applied_index: 95455, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 3247221
E20231002 20:46:02.360119 202086 backward.hpp:4200] Stack trace (most recent call last) in thread 202086:
E20231002 20:46:02.360256 202086 backward.hpp:4200] #11   Object "", at 0xffffffffffffffff, in 
E20231002 20:46:02.360287 202086 backward.hpp:4200] #10   Object "/usr/lib/x86_64-linux-gnu/libc.so.6", at 0x7f99b9f109ff, in 
E20231002 20:46:02.360307 202086 backward.hpp:4200] #9    Object "/usr/lib/x86_64-linux-gnu/libc.so.6", at 0x7f99b9e7eb42, in 
E20231002 20:46:02.360332 202086 backward.hpp:4200] #8    Object "/usr/bin/typesense-server", at 0x55ff817a85c3, in execute_native_thread_routine
E20231002 20:46:02.360354 202086 backward.hpp:4200] #7  | Source "include/threadpool.h", line 57, in operator()
E20231002 20:46:02.360373 202086 backward.hpp:4200]       Source "/usr/include/c++/10/future", line 1592, in ThreadPool [0x55ff7f64f23c]
E20231002 20:46:02.360390 202086 backward.hpp:4200] #6  | Source "/usr/include/c++/10/future", line 1459, in _M_set_result
E20231002 20:46:02.360416 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/future", line 412, in call_once<void (std::__future_base::_State_baseV2::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>()>*, bool*), std::__future_base::_State_baseV2*, std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>()>*, bool*>
E20231002 20:46:02.360440 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/mutex", line 729, in __gthread_once
E20231002 20:46:02.360457 202086 backward.hpp:4200]       Source "/usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h", line 700, in _M_run [0x55ff7f73f0c3]
E20231002 20:46:02.360479 202086 backward.hpp:4200] #5    Object "/usr/lib/x86_64-linux-gnu/libc.so.6", at 0x7f99b9e83f67, in 
E20231002 20:46:02.360503 202086 backward.hpp:4200] #4  | Source "/usr/include/c++/10/future", line 572, in operator()
E20231002 20:46:02.360522 202086 backward.hpp:4200]       Source "/usr/include/c++/10/bits/std_function.h", line 622, in _M_do_set [0x55ff7f64e062]
E20231002 20:46:02.360546 202086 backward.hpp:4200] #3  | Source "/usr/include/c++/10/bits/std_function.h", line 292, in __invoke_r<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<_Fn, _Alloc, _Res(_Args ...)>::_M_run<std::_Bind<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>()>, std::allocator<int>, void, {}>::<lambda()>, void>&>
E20231002 20:46:02.360572 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 115, in __invoke_impl<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<_Fn, _Alloc, _Res(_Args ...)>::_M_run<std::_Bind<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>()>, std::allocator<int>, void, {}>::<lambda()>, void>&>
E20231002 20:46:02.360622 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 60, in operator()
E20231002 20:46:02.360648 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/future", line 1397, in operator()
E20231002 20:46:02.360672 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/future", line 1456, in __invoke_r<void, std::_Bind<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>()>&>
E20231002 20:46:02.360692 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 110, in __invoke_impl<void, std::_Bind<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>()>&>
E20231002 20:46:02.360713 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 60, in operator()<>
E20231002 20:46:02.360731 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/functional", line 499, in __call<void>
E20231002 20:46:02.360749 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/functional", line 416, in __invoke<HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>&>
E20231002 20:46:02.360770 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 95, in __invoke_impl<void, HttpServer::process_request(const std::shared_ptr<http_req>&, const std::shared_ptr<http_res>&, route_path*, const h2o_custom_req_handler_t*, bool)::<lambda()>&>
E20231002 20:46:02.360793 202086 backward.hpp:4200]     | Source "/usr/include/c++/10/bits/invoke.h", line 60, in operator()
E20231002 20:46:02.360811 202086 backward.hpp:4200]       Source "src/http_server.cpp", line 684, in _M_invoke [0x55ff7f740e23]
E20231002 20:46:02.360832 202086 backward.hpp:4200] #2    Source "src/core_api.cpp", line 403, in get_search [0x55ff7f7075cc]
E20231002 20:46:02.360855 202086 backward.hpp:4200] #1  | Source "external/lrucache/include/lru/internal/base-cache.hpp", line 913, in _insert_new_key
E20231002 20:46:02.360872 202086 backward.hpp:4200]     | Source "external/lrucache/include/lru/internal/base-cache.hpp", line 1544, in _evict_lru_for
E20231002 20:46:02.360894 202086 backward.hpp:4200]       Source "external/lrucache/include/lru/internal/base-cache.hpp", line 1558, in insert [0x55ff7f71bd26]
E20231002 20:46:02.360916 202086 backward.hpp:4200] #0    Object "/usr/bin/typesense-server", at 0x55ff817292d1, in std::__detail::_List_node_base::_M_transfer(std::__detail::_List_node_base*, std::__detail::_List_node_base*)
I20231002 20:46:02.406693 202401 batched_indexer.cpp:285] Running GC for aborted requests, req map size: 0
I20231002 20:46:05.902396 202400 raft_server.cpp:564] Term: 4, last_index index: 95455, committed_index: 95455, known_applied_index: 95455, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 3247221
E20231002 20:46:09.718156 202086 typesense_server.cpp:121] Typesense 0.25.1 is terminating abruptly.
Log file created at: 2023/10/02 20:46:10
```

## Expected Behavior

Don't crash

## Actual Behavior

Crashing 

## Metadata

**Typesense Version**: 

0.25.1

**OS**:
 Ubuntu 22.04.3 LTS (GNU/Linux 5.15.0-84-generic x86_64)

**Comment by kishorenc** [comment]
@bonsempregos 

Does this happen repeatedly often or a one-off crash? Can you also verify that the node has enough memory?

**Comment by bonsempregos** [comment]
@kishorenc 

Was a one-off crash, I think it could be something related to cache. as I have a 5s cache on client side. Yes every node have 64GB available.

**Comment by kishorenc** [comment]
@bonsempregos 

Found a potential issue which is fixed in `0.25.2.rc4`.

**Comment by bonsempregos** [comment]
Thanks a lot @kishorenc, just deployed on cluster, if something wrong I will let you know.
